core/src/main/scala/kafka/log/OffsetMap.scala

```diff
@@ -125,10 +143,47 @@ class SkimpyOffsetMap(val memory: Int, val hashAlgorithm: String = "MD5") extend
         return -1L
       bytes.get(hash2)
       attempt += 1
+      if (attempt > slots)
+        return -1L
     } while(!Arrays.equals(hash1, hash2))
     bytes.getLong()
   }
-  
+
+  /**
+    * This copies all entries from an OffsetMap, taking all or none of the entries depending on available space. Values
+    * for identical keys are overwritten.
+    *
+    * @param from The OffsetMap to copy from
+    *
+    * @return If the put succeeded.
+    */
+  override def putAll(from: OffsetMap): Boolean = {
+    require(from.getClass == this.getClass, "cannot merge %s with SkimpyOffsetMap".format(from.getClass))
+
+    val map = from.asInstanceOf[SkimpyOffsetMap]
+    require(map.hashAlgorithm == hashAlgorithm, "SkimpyOffsetMap cannot merge maps with different hashAlgorithms")
```

Comment:
Is this a reasonable restriction? Can we not fall back to a less efficient algorithm if `from` is not a `SkimpyOffsetMap`?